WordPress/src/main/java/org/wordpress/android/models/networkresource/ListState.kt

```diff
@@ -0,0 +1,125 @@
+package org.wordpress.android.models.networkresource
+
+/**
+ * ListState aims to give a highly structured and easy to use way to manage any list's state.
+ *
+ * There are 5 different states: [Init], [Ready], [Success], [Loading], [Error]. Check out their documentation to see
+ * how each state behaves.
+ *
+ * @property data is initialized depending on each state and once initialized it can not be altered. In [Ready] and
+ * [Success] states, it'll be passed as a parameter. In [Loading] and [Error] states, it'll be initialized from the
+ * previous state to make sure the access to the data is not lost. In [Init], an empty list will be passed.
+ * In situations where the data needs to be changed outside of a fetch [getTransformedListState] can be used
+ * to get a new instance by using a transform function.
+ */
+sealed class ListState<T>(val data: List<T>) {
+    /**
+     * In some situations the underlying data might change outside of a fetch. Adding a new item, removing one,
+     * a single item getting updated would be some typical examples. Since the [data] property can not be altered
+     * directly, by design, we need a different way to update it.
+     *
+     * This method can be used to handle any transformation easily while preserving the current state. Any function
+     * that takes a [List] and returns a new one can be used as the transformation.
+     *
+     * @return a new ListState instance that has the transformed data while preserving the state
+     */
+    abstract fun getTransformedListState(transform: (List<T>) -> List<T>): ListState<T>
```

Comment:
I don't think `mutate` is correct here, because that implies that the current instance will change and that goes directly against the `immutable` idea. Having said that, I agree `getTransformedListState` doesn't feel right either. Maybe `transformedListState` would be better? That way we don't have a `getter` type method, it indicates that the state will transform and having the `listState` suffix will hopefully mean that a new instance will be returned. What do you think?